File: main/java/com/websarva/wings/android/dowasurememo/SizeMemo.java
package com.websarva.wings.android.dowasurememo;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.database.CursorIndexOutOfBoundsException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteStatement;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;

public class SizeMemo extends AppCompatActivity {

    private Databasehelper _helper;
    private String _category = "size";
    Context context = SizeMemo.this;
    String table = "size";

    String strBodyPart;
    String strRecord;
    String strUnit;

    EditText etBodyPart;
    EditText etRecord;
    EditText etUnit;
    ImageButton btDelete;

    LinearLayout llSizeLayout;
    LayoutInflater inflater;
    LinearLayout llSizeInputform;

    private int indexCounter = 2;
    int tagId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_size_memo);

        Intent intent = getIntent();
        llSizeLayout = findViewById(R.id.ll_size_layout);

//        llSizeLayout.removeAllViews();
//        DatabaseControl control4=new DatabaseControl(context,table);
//        control4.DatabaseAllDelete();

        _helper = new Databasehelper(getApplicationContext());
        SQLiteDatabase db = _helper.getWritableDatabase();
        String sqlSelect = "SELECT * FROM size";
        Cursor cursor = db.rawQuery(sqlSelect, null);
        cursor.moveToFirst();
        while (cursor.moveToNext()) {
            int i = cursor.getColumnIndex("_id");
            tagId = cursor.getInt(i);
            Log.d("oncreate66",""+tagId);

            inflater = LayoutInflater.from(getApplicationContext());
            llSizeInputform = (LinearLayout) inflater.inflate(R.layout.size_inputform, null);
            llSizeLayout.addView(llSizeInputform);
            ImageView circle = llSizeInputform.findViewById(R.id.circle);
            circle.setColorFilter(Color.rgb(127, 255, 212));

            etBodyPart = llSizeInputform.findViewById(R.id.et_bodypart);
            etRecord = llSizeInputform.findViewById(R.id.et_record);
            etUnit = llSizeInputform.findViewById(R.id.et_unit);
            btDelete = llSizeInputform.findViewById(R.id.bt_delete);
            btDelete.setOnClickListener
                    (new DeleteButton(context,llSizeLayout,llSizeInputform,table));

            i = cursor.getColumnIndex("bodypart");
            strBodyPart = cursor.getString(i);

            i = cursor.getColumnIndex("records");
            strRecord = cursor.getString(i);

            i = cursor.getColumnIndex("unit");
            strUnit = cursor.getString(i);


            try {
                etBodyPart.setText(strBodyPart);
            } catch (NullPointerException e) {
                strBodyPart = "";
            }

            try {
                etRecord.setText(strRecord);
            } catch (NullPointerException e) {
                strRecord = "";
            }

            try {
                etUnit.setText(strUnit);
            } catch (NullPointerException e) {
                strUnit = "";
            }

        }

        if(llSizeLayout.getChildCount()!=0){
            LinearLayout firstView= (LinearLayout) llSizeLayout.getChildAt(0);
            firstView.setVisibility(View.GONE);
        }else {
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.optionmenu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.option_add:
                if(llSizeLayout.getChildCount()==0){
                    inflater = LayoutInflater.from(getApplicationContext());
                    llSizeInputform=(LinearLayout)inflater.inflate(R.layout.size_inputform,null);
                    llSizeLayout.addView(llSizeInputform);
                    llSizeInputform.setVisibility(View.GONE);

                    String str="";
                    DatabaseControl control = new DatabaseControl(context, table);
                    control.DatabaseDelete(1);

                    String column1 = "bodypart";
                    String column2 = "records";
                    String column3 = "unit";

                    DatabaseControl control2 = new DatabaseControl
                            (context, table,1, _category, str, str, str);
                    control2.DatabaseInsertThreeColumns(column1, column2, column3);
                }

                llSizeLayout=findViewById(R.id.ll_size_layout);
                inflater = LayoutInflater.from(getApplicationContext());
                llSizeInputform = (LinearLayout) inflater.inflate(R.layout.size_inputform, null);
                llSizeLayout.addView(llSizeInputform);

                ImageView circle = llSizeInputform.findViewById(R.id.circle);
                circle.setColorFilter(Color.rgb(127, 255, 212));

                etBodyPart = llSizeInputform.findViewById(R.id.et_bodypart);
                etRecord = llSizeInputform.findViewById(R.id.et_record);
                etUnit = llSizeInputform.findViewById(R.id.et_unit);
                btDelete = llSizeInputform.findViewById(R.id.bt_delete);
                btDelete.setOnClickListener
                        (new DeleteButton(context,llSizeLayout,llSizeInputform,table));
        }
        return super.onOptionsItemSelected(item);
    }


    @Override
    protected void onPause() {
        super.onPause();

        indexCounter = 2;

        for (int i = 0; i < llSizeLayout.getChildCount(); i++) {
            LinearLayout linearLayout = (LinearLayout) llSizeLayout.getChildAt(i);
            etBodyPart = linearLayout.findViewById(R.id.et_bodypart);
            etRecord = linearLayout.findViewById(R.id.et_record);
            etUnit = linearLayout.findViewById(R.id.et_unit);

            strBodyPart = etBodyPart.getText().toString();
            strRecord = etRecord.getText().toString();
            strUnit = etUnit.getText().toString();

            DatabaseControl control = new DatabaseControl(context, table);
            control.DatabaseDelete(indexCounter);

            String column1 = "bodypart";
            String column2 = "records";
            String column3 = "unit";

            DatabaseControl control2 = new DatabaseControl
                    (context, table, indexCounter, _category, strBodyPart, strRecord, strUnit);
            control2.DatabaseInsertThreeColumns(column1, column2, column3);

            Log.d("pause194", "" + indexCounter);
            indexCounter++;

        }
    }
}

File: main/java/com/websarva/wings/android/dowasurememo/SizeMemo.kt
package com.websarva.wings.android.dowasurememo

import android.content.Context
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.widget.*
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity

/**
 * 体のサイズを入力するSizeメモのクラス
 *
 * @author nakayama
 * @version 1.0
 */
class SizeMemo : AppCompatActivity() {
    private val context: Context = this@SizeMemo
    private var inflater: LayoutInflater? = null
    private var llSizeLayout: LinearLayout? = null
    private var llSizeInputform: LinearLayout? = null
    private var etBodyPart: EditText? = null
    private var etRecord: EditText? = null
    private var etUnit: EditText? = null
    private var btDelete: ImageButton? = null
    private var strBodyPart: String? = null
    private var strRecord: String? = null
    private var strUnit: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_size_memo)
        inflater = LayoutInflater.from(applicationContext)
        llSizeLayout = findViewById(R.id.ll_size_layout)
        llSizeInputform = inflater.inflate(R.layout.size_inputform, null) as LinearLayout
        /**
         * データベースの列名の配列
         */
        val columnNames = listOf("bodypart", "records", "unit")

        //データベースからデータを取り出して、レイアウトを作成する処理
        val control = DatabaseControl(context, TABLE, columnNames)
        control.selectDatabase(llSizeLayout)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.optionmenu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.option_add -> {
                inflater = LayoutInflater.from(applicationContext)
                llSizeInputform = inflater.inflate(R.layout.size_inputform, null) as LinearLayout
                llSizeLayout!!.addView(llSizeInputform)
                val circle = llSizeInputform!!.findViewById<ImageView>(R.id.circle)
                circle.setColorFilter(Color.rgb(127, 255, 212))
                btDelete = llSizeInputform!!.findViewById(R.id.bt_delete)
                btDelete.setOnClickListener(DeleteButton(context, llSizeLayout, llSizeInputform, TABLE))
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPause() {
        super.onPause()

        //データベースにある全てのデータを削除
        val control = DatabaseControl(context, TABLE)
        control.deleteAllDatabase()

        //メモの文字列を取得してデータベースにインサートする
        for (i in 0 until llSizeLayout!!.childCount) {
            val linearLayout = llSizeLayout!!.getChildAt(i) as LinearLayout
            etBodyPart = linearLayout.findViewById(R.id.et_bodypart)
            etRecord = linearLayout.findViewById(R.id.et_record)
            etUnit = linearLayout.findViewById(R.id.et_unit)
            strBodyPart = etBodyPart.getText().toString()
            strRecord = etRecord.getText().toString()
            strUnit = etUnit.getText().toString()
            val control2 = DatabaseControl(context, TABLE, i, _CATEGORY, strBodyPart, strRecord, strUnit)
            control2.insertDatabaseThreeColumns("bodypart", "records", "unit")
        }
        finish()
    }

    companion object {
        private const val _CATEGORY = "SIZE"

        /**
         * データベースのテーブル名
         */
        private const val TABLE = "size"
    }
}